Flake: UT TestStatefulSetControl/CreatePodFailure/Burst/Delete/StatefulSetAutoDeletePVCEnabled

### Which jobs are flaking?

The UT  TestStatefulSetControl/CreatePodFailure/Burst/Delete/StatefulSetAutoDeletePVCEnabled has been flaking with DATARACE
```
{Failed;Failed;  === RUN   TestStatefulSetControl/CreatePodFailure/Burst/Delete/StatefulSetAutoDeletePVCEnabled
W0911 16:13:06.477078   43850 mutation_detector.go:53] Mutation detector is enabled, this will result in memory leakage.
W0911 16:13:06.477172   43850 mutation_detector.go:53] Mutation detector is enabled, this will result in memory leakage.
W0911 16:13:06.477256   43850 mutation_detector.go:53] Mutation detector is enabled, this will result in memory leakage.
W0911 16:13:06.477356   43850 mutation_detector.go:53] Mutation detector is enabled, this will result in memory leakage.
==================
WARNING: DATA RACE
Write at 0x00c000604fa0 by goroutine 1915:
  k8s.io/kubernetes/pkg/controller/statefulset.(*requestTracker).reset()
      /home/prow/go/src/github.com/kubernetes/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/statefulset/stateful_set_control_test.go:2427 +0xac
  k8s.io/kubernetes/pkg/controller/statefulset.(*fakeObjectManager).CreatePod.func2()
      /home/prow/go/src/github.com/kubernetes/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/statefulset/stateful_set_control_test.go:2476 +0x58
  runtime.deferreturn()
      /usr/local/go/src/runtime/panic.go:477 +0x40
  k8s.io/kubernetes/pkg/controller/statefulset.(*StatefulPodControl).CreateStatefulPod()
      /home/prow/go/src/github.com/kubernetes/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/statefulset/stateful_pod_control.go:122 +0xac
........
......
......
<Stack Trace>
      /home/prow/go/src/github.com/kubernetes/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/statefulset/stateful_set_control.go:295 +0x6c

Goroutine 1915 (running) created at:
  k8s.io/kubernetes/pkg/controller/statefulset.slowStartBatch()
      /home/prow/go/src/github.com/kubernetes/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/statefulset/stateful_set_control.go:289 +0x1b4
<GO ROUTINE TRACE>
-------
-----
-------
      /usr/local/go/src/testing/testing.go:1595 +0x1b8
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1648 +0x64
==================
    testing.go:1465: race detected during execution of test
    --- FAIL: TestStatefulSetControl/CreatePodFailure/Burst/Delete/StatefulSetAutoDeletePVCEnabled (0.07s)
;=== RUN   TestStatefulSetControl
    testing.go:1465: race detected during execution of test
--- FAIL: TestStatefulSetControl (13.96s)
;}
```
Sample failing job: https://prow.ppc64le-cloud.cis.ibm.net/view/gs/ppc64le-kubernetes/logs/periodic-kubernetes-unit-test-ppc64le/1701264349017935872

### Which tests are flaking?

TestStatefulSetControl/CreatePodFailure/Burst/Delete/StatefulSetAutoDeletePVCEnabled

### Since when has it been flaking?

Not sure start of flake

### Testgrid link

https://testgrid.k8s.io/ibm-unit-tests-ppc64le

### Reason for failure (if possible)

_No response_

### Anything else we need to know?

_No response_

### Relevant SIG(s)

/sig testing

--- a/pkg/controller/statefulset/stateful_set_control_test.go
+++ b/pkg/controller/statefulset/stateful_set_control_test.go
@@ -2429,6 +2429,12 @@ func (rt *requestTracker) reset() {
 	rt.delay = 0
 }
 
+func (rt *requestTracker) getErr() error {
+	rt.Lock()
+	defer rt.Unlock()
+	return rt.err
+}
+
 func newRequestTracker(requests int, err error, after int) requestTracker {
 	return requestTracker{
 		requests: requests,
@@ -2474,9 +2480,7 @@ func (om *fakeObjectManager) CreatePod(ctx context.Context, pod *v1.Pod) error {
 	defer om.createPodTracker.inc()
 	if om.createPodTracker.errorReady() {
 		defer om.createPodTracker.reset()
-		defer om.createPodTracker.Unlock()
-		om.createPodTracker.Lock()
-		return om.createPodTracker.err
+		return om.createPodTracker.getErr()
 	}
 	pod.SetUID(types.UID(pod.Name + "-uid"))
 	return om.podsIndexer.Update(pod)
@@ -2494,7 +2498,7 @@ func (om *fakeObjectManager) DeletePod(pod *v1.Pod) error {
 	defer om.deletePodTracker.inc()
 	if om.deletePodTracker.errorReady() {
 		defer om.deletePodTracker.reset()
-		return om.deletePodTracker.err
+		return om.deletePodTracker.getErr()
 	}
 	if key, err := controller.KeyFunc(pod); err != nil {
 		return err

--- a/pkg/controller/statefulset/stateful_set_control_test.go
+++ b/pkg/controller/statefulset/stateful_set_control_test.go
@@ -2474,6 +2474,8 @@ func (om *fakeObjectManager) CreatePod(ctx context.Context, pod *v1.Pod) error {
 	defer om.createPodTracker.inc()
 	if om.createPodTracker.errorReady() {
 		defer om.createPodTracker.reset()
+		defer om.createPodTracker.Unlock()
+		om.createPodTracker.Lock()
 		return om.createPodTracker.err
 	}
 	pod.SetUID(types.UID(pod.Name + "-uid"))

